### ISSUE TITLE
First Impressions with V2

### ISSUE THREAD

**Issue by dantleech** [thephpleague/flysystem#1245]
Hey, first of all, version 2 looks really great!

I'm trying to [implement it](https://github.com/dantleech/maestro2/pull/3) on a project which may have more low-level requirements than are typical and noted the following:

- There seems not to be an API to determine if a single location is a file or a directory - only `listContents` provides the attribute classes with this info.
- `fileExists` in the local adapter only works with files (`is_file`), and there is no `directoryExists` method (I can see it was initially using `file_exists` and was changed to `is_file`)
- This application works with literal Unix permissions, would it be worth bundling a 1-1 `UnixPermissionConverter` or similar?

**Comment by frankdejonge** [comment]
Hi @dantleech,

Thanks for providing your feedback. The directory existence was something that was impossible to get sorted out in a cross adapter way. Therefor I settled on only accounting for the thing that was cross-fs compatible, the file existance check. The `has` call in V1 had its own quicks. With all directories being implicitly created, the need for explicitly creating directories does not exist. These things combined makes that the API is more focused around files than directories.

Is there a particular situation you're struggling with, perhaps I can help out from my perspective.

The visibility converters allow you to create a mapping between any string input and whatever you want to convert it too/from. So this means if you have the need to create a more direct mapping, this is of course possible. If you want to contribute one, I'm happy to review and release it.

**Comment by dantleech** [comment]
It may be that this application isn't the best use case for Flysystem - it's a repository orchestration system (similar to Ansible) so may have some peculiar requirements:

- `->fileOrDirectoryExists()`: filtering out paths which do not exist (e.g. filtering which paths are passed to `git add`)
- `->isDirectory()` The application can say that `file` should exist at a given path, and that file currently existing as a directory, then an error should be thrown.

As an extra I also needed to expose the absolute "host" path for the filesystem so that it can be used as the CWD for processes etc. 

```
new Process(['composer', 'install'], $filesystem->absoluteResolvedPath('path/relative/to/workspace/root'));
```

For the time being I created a cheap Filesystem abstraction which does the job...

I'll close this issue now also - thanks for your response and offer of help ;)
